lib/spack/spack/package_base.py

```diff
@@ -1667,66 +1666,75 @@ def all_patches(cls):
+                  "sha256": "56ab9b90f5acbc42eb7a94cf482e6c058a63e8a1effdf572b8b2a6323a06d923",
+                  "type": "archive"
+                }
+             }
+
+        All hashes are added to concrete specs at the end of concretization. If this
+        method is called on an abstract spec, only hashes that can be known from the
+        abstract spec will be included.
 
         """
-        # list of components to make up the hash
-        hash_content = []
+        hashes = syaml.syaml_dict()
 
         # source artifacts/repositories
-        # TODO: resources
         if self.spec.versions.concrete:
+            sources = []
             try:
-                source_id = fs.for_package_version(self).source_id()
+                fetcher = fs.for_package_version(self)
+                sources.append(fetcher.spec_attrs())
+
             except (fs.ExtrapolationError, fs.InvalidArgsError):
                 # ExtrapolationError happens if the package has no fetchers defined.
                 # InvalidArgsError happens when there are version directives with args,
                 #     but none of them identifies an actual fetcher.
-                source_id = None
 
-            if not source_id:
-                # TODO? in cases where a digest or source_id isn't available,
-                # should this attempt to download the source and set one? This
-                # probably only happens for source repositories which are
-                # referenced by branch name rather than tag or commit ID.
+                # if this is a develop spec, say so
                 env = spack.environment.active_environment()
                 from_local_sources = env and env.is_develop(self.spec)
+
+                # don't bother setting a hash if none is available, but warn if
+                # it seems like there should be one.
                 if self.has_code and not self.spec.external and not from_local_sources:
-                    message = "Missing a source id for {s.name}@{s.version}"
+                    message = "Missing a hash for {s.name}@{s.version}"
                     tty.debug(message.format(s=self))
-                hash_content.append("".encode("utf-8"))
-            else:
-                hash_content.append(source_id.encode("utf-8"))
+
+            for resource in self._get_needed_resources():
+                sources.append(resource.fetcher.spec_attrs())
+
+            if sources:
+                hashes["sources"] = sources
 
         # patch sha256's
         # Only include these if they've been assigned by the concretizer.
         # We check spec._patches_assigned instead of spec.concrete because
         # we have to call package_hash *before* marking specs concrete
         if self.spec._patches_assigned():
-            hash_content.extend(
-                ":".join((p.sha256, str(p.level))).encode("utf-8") for p in self.spec.patches
-            )
+            hashes["patches"] = [p.sha256 for p in self.spec.patches]
```

Comment:
good point -- added this back.

Reply:
Ok, adding this back breaks spec reading, which is something I had wanted to avoid.  Apparently:

1. We encode `patches` as a variant.
2. We encode `patches` at the top level of the spec dictionary as a list of hashes, and it needs to stay there to be read properly by old versions of Spack.  I *could* bump the spec version, but was hoping not to.  Trying to decide whether we should.
3. We should preserve patch level, but I don't think that it can go in either of the other two places. The place to put it is *probably* in the sources list, along with other artifacts. I don't really like this b/c now `patches` will appear in 3 places on the spec, but maybe the others can be deprecated.  Maybe this is a good enough reason to bump the spec version, or at least something we should do when we next bump it.